### PR TITLE
fix Issue 17582 - Applying const to struct declaration should make th…

### PR DESCRIPTION
**Pull request by WalterBright**
…e struct type const

### CODE REVIEW

**Comment by UplinkCoder** [comment]
do we really need this ?
I supposed it does no harm though.

**Comment by WalterBright** [comment]
See also https://github.com/dlang/dmd/pull/783 where this bug was introduced.

**Comment by UplinkCoder** [comment]
@WalterBright I think you can merge this.
There is no harm here.

File: test/runnable/testconst.d [hunk 1]
```diff
@@ -2806,9 +2806,9 @@ void test6982()
 /************************************/
 // 7038
```
Comment (andralex) [inline test/runnable/testconst.d]: Just to really nail it down, could you please write this:
```d
static assert(is(S7038 == const S7038)));
```

There should be tests for the other qualifiers as well, notably shared.

**Review by andralex — CHANGES_REQUESTED** [review]

**Comment by andralex** [comment]
This is a change in language semantics so it needs a changelog.

I also realized the following problem. Consider:

```d
immutable struct S
{
    void fun() {}
    void fun() immutable {}
}
```

This should not compile because fun is the same function defined twice. Similarly:

```d
immutable struct S {}
void fun(S) {}
void fun(immutable S) {}
```

should not work because fun is defined twice.

**Comment by WalterBright** [comment]
> This should not compile

It's actually an old issue.
```
void foo() { }
void foo() { }
```
compiles but does not link, as does your example. I.e. it is a separate issue from this PR.

File: test/runnable/testconst.d (hunk 1 above)
Comment (WalterBright) [inline test/runnable/testconst.d]: Since the code to implement this is generic for all qualifiers, that is unnecessary. But I'll add one for shared.

**Review by andralex — APPROVED** [review]
Thx! Can you please add to the changelog before merging?

**Comment by andralex** [comment]
The next big step is classes and interfaces.

**Comment by UplinkCoder** [comment]
is `inout struct` something that even makes sense ?

**Comment by WalterBright** [comment]
> Can you please add to the changelog before merging?

Since it's a regression fix it will anyway.

**Comment by wilzbach** [comment]
> Since it's a regression fix it will anyway.

Yup, but as you have already recognized this as regression - why didn't you base it onto stable?
Imho one needs to try really hard to overlook this warning:

image

CC @CyberShadow - I am running out of ideas here:
- should we call it "Errors"?
- should we block auto-merge?

**Comment by CyberShadow** [comment]
The PR was created before the new warning notice was added, so let's give it a bit more time.

**Comment by ZombineDev** [comment]
@wilzbach the regression is from 2.059. It would be too risky to include this fix in stable, IMO.

**Pull request merged**